### PR TITLE
CMake, Doxygen Documentation, GTest Unit Test, and Compiler warning fixes.

### PR DESCRIPTION
**Pull request by mcsauder**
Hello,

I've tried to help out a bit by creating this pull request which adds CMake files, GTest unit testing, a lot of doxy style documentation in the header files, and corrected a lot of "const" related compiler warnings.

No structural changes to function names or anything that would break backwards compatibility is in this pull request.

If you would like to see anything else accomplished or have concerns over the PR please let me know!

-Mark Sauder

### CODE REVIEW

**Comment by crayzeewulf** [comment]
Mark:

Thanks a lot for the updates via this pull request. I will review and merge shortly.

Cheers,
--
CrayzeeWulf

**Comment by mcsauder** [comment]
I've updated the GTest URL in the Top level CMakeLists.txt.  If you are aware of how to structure that CMakeLists.txt to locate the installed gtest library instead of downloading the zip and building it, please let me know.

Aside from being able to improve the CMake Gtest dependency, I think everything is in good working order, I've cloned a clean repo and run the compile script, and everything compiles fine.  As long as you can physically connect two serial ports (they have to be named correctly in the unit test as well), all unit tests pass!

Please let me know if you have any questions.

-Mark

**Comment by crayzeewulf** [comment]
Thanks again, Mark. I use CMake quite frequently so I ought to be able to figure out how to detect an existing installation of GTest (maybe using an environment variable as a hint). I will try it out after I merge your changes. 
--
CrayzeeWulf

**Comment by crayzeewulf** [comment]
Mark:

I am going through the various changes in the pull request and will add comments as I proceed. Here is the first set of comments.

There is a specific reason why there were several headers included in the `.cpp` files and not in the `.h` files (for example, `<cerrno>`, `<cassert>`, etc. in `SerialPort.cpp` but not in `SerialPort.h`). They are needed only in the implementation of the classes and users of these classes do not need them. So, someone who needs to use `SerialPort` class does not need to include these headers. I recommend that these headers be moved back to the `.cpp` files. This applies to the following files in the pull request:

- `PosixSignalDispatcher.cpp` and `PosixSignalDispatcher.h`
- `SerialPort.cpp` and `SerialPort.h`
- `SerialStream.cc` and `SerialStream.h`
- `SerialStreamBuf.cc` and `SerialStreamBuf.h`

Basically, the `.h` files should only contain the minimum set of headers that are needed to compile the header file by itself. The rest should be in `.cpp` file because they are an implementation detail that users of the class should not see. This avoids unnecessary dependencies of users' code (which uses our headers) on headers that they do not need. 

See these discussions for example:

- http://programmers.stackexchange.com/a/167751
- Guideline#9 in http://umich.edu/~eecs381/handouts/CppHeaderFileGuidelines.pdf

Could you please move the `#include` statements back into the `.cpp` and `.cc` files ?

**Comment by crayzeewulf** [comment]
Regarding changes to `doxygen.conf.in`, I see that you replaced `@PACKAGE@` and `@VERSION@` with fixed values. 

Note that the build based on GNU autotools automatically generates `doxygen.conf` file from `doxygen.conf.in` and fills in the correct package name and version number in place of `@PACKAGE@` and `@VERSION@`. The actual package name and version number are defined at one place at the top of `configure.ac` file. So, the `@PACAKGE@` and `@VERSION@` should be left as is in the `doxygen.conf.in` file. 

Similar functionality is available in CMake too via [`configure_file`](https://cmake.org/cmake/help/v3.0/command/configure_file.html). For a simple example, see [this](https://tty1.net/blog/2014/cmake-doxygen_en.html) for generating `doxygen.conf` file with some fields filled in.

**Comment by mcsauder** [comment]
Hi Crayzeewulf,

I've reverted the doxyfile to the original content and the #includes to
their original locations.  Thanks for your feedback on that!

The ReadLine() call in the unit test is still commented out as it does not
pass, but all others are being tested and pass.

I didn't realize that the doxyfile was autogenerated.  I actually couldn't
even get the library to build from your repo master branch until just two
days ago, I had a hard time figuring out what packages were missing and it
wasn't very straightforward from the documentation to me.  That was
actually the reason I put the CMake files together to begin with, along
with the GTest framework I worked at instituting.

In case it is helpful at all, I would be happy to decline the current pull
request and create a new one from a single commit so that things show up a
little cleaner.

Additionally, if it's helpful, I could submit a smaller subset of this work
in multiple pull requests.

Last, feel free to simply checkout a clone of my repo and copy/paste from
my files if you prefer bringing changes in like that instead.

At this point, I think the main things the PR offers are doxy comments,
CMake, and the unit test file.

Let me know if you'd like to see anything else done differently and I'm
happy to get to work!  Thanks for your feedback and suggestions!

Respectfully,

Mark Sauder, M.S. M.E.
CEO, *Grey Point Corporation* <http://www.greypointcorp.com/>
c: 801.698.8786
mcsauder@greypointcorp.com

**Comment by mcsauder** [comment]
This PR has gotten a little bit messy, even though I am still happy with what was accomplished.  I will close this PR and open a new one that has fewer differences but captures the work that got done.

-Mark
